pkg/pipeline/extract/conntrack/conntrack.go

```diff
@@ -116,65 +116,50 @@ func (ct *conntrackImpl) Extract(flowLogs []config.GenericMap) []config.GenericM
-	ct.connStore.iterateFrontToBack(expiryOrder, func(conn connection) (shouldDelete, shouldStop bool) {
-		expiryTime := conn.getExpiryTime()
-		if ct.clock.Now().After(expiryTime) {
-			// The connection has expired. We want to pop it.
-			record := conn.toGenericMap()
-			addHashField(record, conn.getHash().hashTotal)
-			addTypeField(record, api.ConnTrackOutputRecordTypeName("EndConnection"))
-			var isFirst bool
-			if ct.shouldOutputEndConnection {
-				isFirst = conn.markReported()
-			}
-			addIsFirstField(record, isFirst)
-			outputRecords = append(outputRecords, record)
-			shouldDelete, shouldStop = true, false
-		} else {
-			// No more expired connections
-			shouldDelete, shouldStop = false, true
+	// Convert the connections to GenericMaps and add meta fields
+	for _, conn := range connections {
+		record := conn.toGenericMap()
+		addHashField(record, conn.getHash().hashTotal)
+		addTypeField(record, api.ConnTrackOutputRecordTypeName("EndConnection"))
+		var isFirst bool
+		if ct.shouldOutputEndConnection {
+			isFirst = conn.markReported()
 		}
-		return
-	})
+		addIsFirstField(record, isFirst)
+		outputRecords = append(outputRecords, record)
+	}
 	return outputRecords
 }
 
 func (ct *conntrackImpl) prepareUpdateConnectionRecords() []config.GenericMap {
+	connections := ct.connStore.prepareUpdateConnections()
+
 	var outputRecords []config.GenericMap
-	// Iterate over the connections by their next update report time from old to new.
-	ct.connStore.iterateFrontToBack(nextUpdateReportTimeOrder, func(conn connection) (shouldDelete, shouldStop bool) {
-		nextUpdate := conn.getNextUpdateReportTime()
-		needToReport := ct.clock.Now().After(nextUpdate)
-		if needToReport {
-			record := conn.toGenericMap()
-			addHashField(record, conn.getHash().hashTotal)
-			addTypeField(record, api.ConnTrackOutputRecordTypeName("UpdateConnection"))
-			var isFirst bool
-			if ct.shouldOutputUpdateConnection {
-				isFirst = conn.markReported()
-			}
-			addIsFirstField(record, isFirst)
-			outputRecords = append(outputRecords, record)
-			newNextUpdate := ct.clock.Now().Add(ct.config.UpdateConnectionInterval.Duration)
-			ct.connStore.updateNextReportTime(conn.getHash().hashTotal, newNextUpdate)
-			shouldDelete, shouldStop = false, false
-		} else {
-			shouldDelete, shouldStop = false, true
+	// Convert the connections to GenericMaps and add meta fields
+	for _, conn := range connections {
+		record := conn.toGenericMap()
+		addHashField(record, conn.getHash().hashTotal)
+		addTypeField(record, api.ConnTrackOutputRecordTypeName("UpdateConnection"))
+		var isFirst bool
+		if ct.shouldOutputUpdateConnection {
+			isFirst = conn.markReported()
 		}
-		return
-	})
+		addIsFirstField(record, isFirst)
+		outputRecords = append(outputRecords, record)
+	}
 	return outputRecords
 }
 
+// TBD: think of how to avoid confusion with the word "update". It is used with periodic connections updates and for update a connection with incoming flow logs.
```

Comment:
Just a suggestion, you could use the terminology of "heartbeats" to refer to the periodic updates on long connections?

Reply:
I like this idea! thanks @jotak :)